src/PointwiseFunctions/GeneralRelativity/WeylPropagating.cpp

```diff
@@ -0,0 +1,130 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/GeneralRelativity/WeylPropagating.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "PointwiseFunctions/GeneralRelativity/WeylElectric.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+
+/// \cond
+namespace gr {
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::ii<DataType, SpatialDim, Frame> weyl_propagating(
+    const tnsr::ii<DataType, SpatialDim, Frame>& ricci,
+    const tnsr::ii<DataType, SpatialDim, Frame>& extrinsic_curvature,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::ijj<DataType, SpatialDim, Frame>& cov_deriv_extrinsic_curvature,
+    const tnsr::I<DataType, SpatialDim, Frame>& unit_interface_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& projection_IJ,
+    const tnsr::ii<DataType, SpatialDim, Frame>& projection_ij,
+    const tnsr::Ij<DataType, SpatialDim, Frame>& projection_Ij,
+    const short sign) noexcept {
+  tnsr::ii<DataType, SpatialDim, Frame> weyl_prop{};
+  weyl_propagating<SpatialDim, Frame, DataType>(
+      make_not_null(&weyl_prop), ricci, extrinsic_curvature,
+      inverse_spatial_metric, cov_deriv_extrinsic_curvature,
+      unit_interface_normal_vector, projection_IJ, projection_ij, projection_Ij,
+      sign);
+  return weyl_prop;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void weyl_propagating(
+    const gsl::not_null<tnsr::ii<DataType, SpatialDim, Frame>*> weyl_prop_u8,
+    const tnsr::ii<DataType, SpatialDim, Frame>& ricci,
+    const tnsr::ii<DataType, SpatialDim, Frame>& extrinsic_curvature,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::ijj<DataType, SpatialDim, Frame>& cov_deriv_extrinsic_curvature,
+    const tnsr::I<DataType, SpatialDim, Frame>& unit_interface_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& projection_IJ,
+    const tnsr::ii<DataType, SpatialDim, Frame>& projection_ij,
+    const tnsr::Ij<DataType, SpatialDim, Frame>& projection_Ij,
+    const short sign) noexcept {
+  if (UNLIKELY(get_size(get<0, 0>(*weyl_prop_u8)) !=
+               get_size(get<0>(unit_interface_normal_vector)))) {
+    *weyl_prop_u8 = tnsr::ii<DataType, SpatialDim, Frame>(
+        get_size(get<0>(unit_interface_normal_vector)));
+  }
+
+  auto temp =
```

Comment:
maybe use a `Variables` and `not_null` to reduce to 1 allocation instead of 6?

Reply:
I'm not sure how would allocating memory and passing a pointer to it to `gr::weyl_electric` help reduce the number of allocations. Maybe you mean something that I'm not seeing here?

Reply:
`weyl_electric` is currently allocating a `tnsr::ii<DataVector,...>` which does 6 allocations. If a `not_null` of a `Variables`-allocated tensor of the same rank is passed in, then you don't do the 6 allocations, you just do 1 for the Variables. Does that make sense? Maybe I've misunderstood the code too :)

Reply:
I see, I thought you meant a way to reduce the size of allocations and not only the number. Have updated the code to using `Variables` through `TempBuffer` (so we can accommodate `double`s)